### ISSUE TITLE
docker exec --privileged no longer sets cap_sys_admin properly

### ISSUE THREAD

**Issue by brendangregg** [docker/docker#32018]
We're running docker containers, and have sometimes wanted to run Linux perf (aka perf_events) from within a container for profiling (flame graphs). Since the perf_event_open syscall is blocked[1], we've typically done this by attaching to the container using:

```docker exec -it --privileged UUID bash```

That recently stopped working. Eg, a ```--privileged``` shell:

```
# perf list sw

List of pre-defined events (to be used in -e):

#
```

That no longer lists events (including the cpu-clock event used for profiling), despite having cap_sys_admin:

```
# capsh --print
Current: = cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,37+eip
Bounding set =cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,37
Securebits: 00/0x0/1'b0
 secure-noroot: no (unlocked)
 secure-no-suid-fixup: no (unlocked)
 secure-keep-caps: no (unlocked)
uid=0(root)
gid=0(root)
groups=
```

If, however, we use ```docker run --cap-add sys_admin```, then a normal shell (```docker attach```):

```
# perf list sw

List of pre-defined events (to be used in -e):

  alignment-faults                                   [Software event]
  bpf-output                                         [Software event]
  context-switches OR cs                             [Software event]
  cpu-clock                                          [Software event]
  cpu-migrations OR migrations                       [Software event]
  dummy                                              [Software event]
  emulation-faults                                   [Software event]
  major-faults                                       [Software event]
  minor-faults                                       [Software event]
  page-faults OR faults                              [Software event]
  task-clock                                         [Software event]
```

and has:

```
# capsh --print
Current: = cap_chown,cap_dac_override,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_net_bind_service,cap_net_raw,cap_sys_chroot,cap_sys_admin,cap_mknod,cap_audit_write,cap_setfcap+eip
Bounding set =cap_chown,cap_dac_override,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_net_bind_service,cap_net_raw,cap_sys_chroot,cap_sys_admin,cap_mknod,cap_audit_write,cap_setfcap
Securebits: 00/0x0/1'b0
 secure-noroot: no (unlocked)
 secure-no-suid-fixup: no (unlocked)
 secure-keep-caps: no (unlocked)
uid=0(root)
gid=0(root)
groups=
```

This is odd since we had cap_sys_admin before when it didn't work, yet running ```--cap-add sys_admin``` makes it work. But we already had that (according to ```capsh```), so that should be a no-op. Is this a bug or intentional behavior?

The old method, of jumping in with ```--privileged``` lets us run containers without the ```--cap-add```. Note that we still need the ```--privileged``` for perf to actually do recording, eg, for this to work fully:

```
perf record -F 99 -a -g -- sleep 10
```

**Steps to reproduce the issue:**

1. echo -1 > /proc/sys/kernel/perf_event_paranoid
2. Create a normal container
3. docker exec bash with --privileged
4. apt-get update; apt-get install -y linux-tools-common linux-tools-`uname -r`
5. perf list sw
6. perf record -F 99 -a -g -- sleep 10

**Reference info**

This is an AWS cloud instance:

```
# docker version
Client:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   78d1802
 Built:        Tue Jan 31 23:35:14 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   78d1802
 Built:        Tue Jan 31 23:35:14 2017
 OS/Arch:      linux/amd64

# docker info
Containers: 20
 Running: 2
 Paused: 0
 Stopped: 18
Images: 1
Server Version: 1.12.6
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null bridge host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.10.0-rc8-virtual
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 14.69 GiB
Name: bgregg-xenial-build-i-xxx
ID: JIDN:FVAR:ABKU:LN7Z:5INL:ONEA:A2DH:LEMH:ZFL5:EMJL:TC3W:VZTH
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

Same issue experienced on a more standard Linux 4.8 kernel.

[1] https://docs.docker.com/engine/security/seccomp/#significant-syscalls-blocked-by-the-default-profile

**Comment by thaJeztah** [comment]
is this the same issue as https://github.com/docker/docker/issues/21984 ?

**Comment by brendangregg** [comment]
thanks @thaJeztah, looks like it is (I missed it as my draft title was "Linux perf no longer works in a container" and hadn't found anything like it).

I'll close for now, since it can be tracked by #21984.

**Comment by thaJeztah** [comment]
No worries, GitHub search isn't great 😄, so not surprised you didn't find it

**Comment by ReSearchITEng** [comment]
adding ` --cap-add sys_admin` solved it on docker 1.10
